service/history/replicationTaskExecutor.go

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination replicationTaskExecutor_mock.go -self_package github.com/uber/cadence/service/history
+
+package history
+
+import (
+	"context"
+
+	"github.com/uber/cadence/.gen/go/history"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/xdc"
+)
+
+type (
+	replicationTaskExecutor interface {
+		execute(sourceCluster string, replicationTask *r.ReplicationTask, forceApply bool) (int, error)
+	}
+
+	replicationTaskExecutorImpl struct {
+		currentCluster      string
+		domainCache         cache.DomainCache
+		nDCHistoryResender  xdc.NDCHistoryResender
+		historyRereplicator xdc.HistoryRereplicator
+		historyEngine       Engine
+
+		metricsClient metrics.Client
+		logger        log.Logger
+	}
+)
+
+func newReplicationTaskExecutor(
```

Comment:
can you add some comments on what / why this class exist ?

Reply:
Added the comment